### ISSUE TITLE
[macOS] Performance issues on Macbook (early 2016)

### ISSUE THREAD

**Issue by tkrajacic** [mapbox/mapbox-gl-native#7820]
When running the latest macOS demo app, there are some performance issues on my Macbook.

I have recorded this using Quicktime player and you can download a mov file here:
[Dropbox Link](https://dl.dropboxusercontent.com/u/4984533/mapbox.mov)

Of course the Macbook doesn't have a lot of power, but what strikes me as odd is that I see heavy CPU usage while zooming and scrolling but almost no GPU usage (I am using iStat menus for this). The CPU is of course a very limiting factor on this machine (only 2 cores iirc).

There are two things to note in the recording:

- I am using the trackpad to very smoothly zooming in and out (The sudden changes in zoom are not because of my sudden behavior)
- When blanketing the map with pins there is lots of flickering. Sometimes this is not only the pins, but also the map flickering white.

Both problems vary in intensity when repeating the test.

/cc @1ec5 

**Platform:** macOS 10.12.2 on Macbook (early 2016)
**Mapbox SDK version:** 0.3.0

### Steps to trigger behavior

 1. Open example app
 2. zoom and scroll around
 3. trigger blanketing of map with pins

### Expected behavior

- Smooth experience

### Actual behavior

- Severe stuttering and sometimes flickering when blanketing map

**Comment by friedbunny** [comment]
I see similar performance issues on my MacBook Pro (Retina, 13-inch, Early 2015).

**Comment by 1ec5** [comment]
> When blanketing the map with pins there is lots of flickering. Sometimes this is not only the pins, but also the map flickering white.

This sounds like #7616, which also affects iOS SDK v3.4.0 and has been fixed on the master branch for v3.5.0 and v0.4.0.

**Comment by 1ec5** [comment]
> I see heavy CPU usage while zooming and scrolling but almost no GPU usage

Sounds like #7834 could be the cause in that case.

**Comment by kkaefer** [comment]
I can also replicate this behavior with the integrated GPU of a quad core Macbook Pro 2016. When rendering locks up, I'm not seeing heavy CPU usage (~2%) though.

**Comment by kkaefer** [comment]
(https://github.com/mapbox/mapbox-gl-native/pull/7834 allows testing the integrated GPU on dual-GPU Macs)

**Comment by kkaefer** [comment]
I tested this with a [patched version of GLFW](https://github.com/glfw/glfw/pull/377#issuecomment-275071403) that supports integrated GPUs, and got flawless, fast rendering with the built-in GPU. This leads me to believe that there's something we're doing wrong when initializing our OpenGL context.

**Comment by 1ec5** [comment]
Probably unlike GLFW, we're rendering onto a layer-backed view, which on macOS means WindowServer has to do compositing for the view. (This enables things like visual effect views and the attribution view, similar to on iOS.) What happens if you turn off layer-backing? (Turn off "full window content view" in NSWindow's Attributes inspector and ensure no views are selected in the Effects inspector.)

**Comment by 1ec5** [comment]
By the way, I just cherry-picked #7834 into the release-ios-v3.4.0 branch (for v0.3.1) as #7861.

**Comment by kkaefer** [comment]
> What happens if you turn off layer-backing?

Same, super sluggish behavior.

**Comment by kkaefer** [comment]
I checked out versions back until June 2016, and applied https://github.com/mapbox/mapbox-gl-native/pull/7834. I got the same abysmal performance on all systems with my integrated GPU.

**Comment by mmase** [comment]
I am also noticing poor performance. I set up the demo app in an effort to explore making a native app instead of a web app for better performance. Was sorely disappointed to find the native app is substantially less performant than running a web app in Chrome.

**Comment by jeeftor** [comment]
Seeing the same

**Comment by ChrisLoer** [comment]
I don't know what to make of this clue, but I think I ran into some variant of this problem while testing #10103. What I saw was that every handful of frames, a call to `MGLMapView setNeedsGLDisplay` would not trigger a following call to `MGLOpenGLLayer drawInOpenGLContext`. If an animation was in progress, that animation would pause until something else triggered a render. The problem went away (and rendering was many times smoother) using GLFW.
